### ISSUE TITLE
Control resolver parameters without indexing twice

### ISSUE THREAD

**Issue by tomersein** [anchore/syft#3007]
**What would you like to be added**:
Hello, today if we want to use squashed\all-layers or use exclusions, we can define it only in one place and one resolver is created for all tasks.
Do you think it is possible to index only once, but still define exclusions for task A, and exclusions for task B, and different scopes (for example - digest to be all-layers and packages to be squashed).
today the only solution is to run the same image\directory twice with different configuration.
**Why is this needed**:
I want to optimize the runtime of syft but still keep the relevant results. for example - I want to exclude node-modules for digest but to include it in packages cataloger.
I want to use squashed for packages, but all layers for metadata\digest.
**Additional context**:

When I tried to look into the code I saw it all happens in the resolver, and it is a shared resource between tasks. maybe it is possible in some way to make few types of scopes\exclusions in the same resolver?

**Comment by wagoodman** [comment]
This idea came up early on, and was even partially implemented (there used to be configuration to have different scopes for different catalogers, which has since been removed), but it was ultimately scrapped to keep the concept for what the SBOM stood for to be as essential as possible. 

That is, when you are looking at the final SBOM, if there are multiple scopes and exclusions for each cataloger then it becomes difficult to reason about what was analyzed and what wasn't.

When it comes to indexing container image, we can create multiple file resolvers cheaply without having to re-index anything. This is because images are already indexed and creating file resolvers is really about selecting one or more file trees to be a part of the resolver. For the directory resolver this is different since it is ultimately not an indexed source, so creating file resolvers incurs the cost of indexing the source. And since exclusion processing is done at the index level of the dir resolver then having different exclusions also incurs that cost.

The image resolvers went a separate direction for exclusions: there is an exclusion decorator that filters results after the fact https://github.com/anchore/syft/blob/c816039e91b46ae475207db01c8eac9fbc139c02/syft/source/stereoscopesource/image_source.go#L114-L117 

> maybe it is possible in some way to make few types of scopes\exclusions in the same resolver?

If we were to implement this, I don't think it would be to add multiple scopes and exclusion options to the same resolver, but instead at a task-level allow tasks to request resolvers based on request parameters (scope and exclusion). This may cause issues when scanning large dirs over images based on the above discussion.

I think whats missing to move this forward is a couple things:

1. What would it look like as a user to configure this? Is it an explicit scope/exclude configurable for each cataloger? is it a new resolver config array that can be attributed to other configuration sections? is it something else?
2. Is this something that would be available at the CLI? or maybe only for advanced usage at the API only? (@tomersein which camp does your needs fall into here?)

**Comment by tomersein** [comment]
Hi!
Thanks for the comprehensive answer.
You are correct regarding the images, since most of the times they are smaller, indexing twice or running twice syft takes less time.
however, when trying to scan large machines it can take a lot of time and memory to index. 
1. As a user I don't think it will matter for me how to configure it, we can do it like it was before (for each cataloger define scope\exclusions).
2. I don't understand the difference between CLI or API in this case, can you elaborate?

the main thing here that I want to exclude paths for md5 cataloger (for example skip all node_moudles) but do not skip it in packages cataloger, and now I can't do it since we index only once. *but* since this action is heavy, I don't want to index it twice..

**Comment by tomersein** [comment]
I think I even have an idea - 
In the code the resolver has some functions of getting files:
`FilesByMIMEType
`
maybe it is possible to `FileByCatalogerExclusions` user can inject -> which will include a map of all files that are relevant for a user

**Comment by TimBrown1611** [comment]
Can be useful if a user uses both files & packages tasks, and don't want to go over all the files in the system for files tasks.
is it possible to support exclusions for all files and exclusions for packages tasks?

**Comment by tomersein** [comment]
hi! I see it was labled "needs discussion"
is it possible to raise it up in the OSS Gardening?
@willmurphyscode

**Comment by willmurphyscode** [comment]
I think the real discussion to have here is: What should we change about Syft to make it better at scanning whole computers, as opposed to the directory scan (which is sort of designed with a local checkout of some project's source code in mind) or an image scan (which is designed with an OCI image in mind).

I'll start a thread about this on Discourse, since I don't think it's really about this particular question. Does that make sense @tomersein?

**Comment by willmurphyscode** [comment]
I'm trying to capture a core, actionable request here so that maybe we can work on it (since I distracted us above :) ).

I think the ask is:

1. It would be nice to have a squashed view of the packages in an image, since this captures the set of packages that are present in the final image.
2. It would be nice to have a layered view of file digests, since this gives insight into how the image was built and what files might be present in it anywhere
3. It would be really nice if I could have 1 and 2 without running Syft twice

@tomersein have I summarized this issue correctly? Can you confirm that you're only asking for this feature for image scans? You mentioned directories and I mentioned VMs above, but only image scans have any notion of layers, so we can't have item 2 on directory or VM sources.
